lib/compress/zstd_compress.c

```diff
@@ -4484,6 +4489,27 @@ static void ZSTD_overflowCorrectIfNeeded(ZSTD_matchState_t* ms,
     }
 }
 
+#include "zstd_preSplit.h"
+
+static size_t ZSTD_optimalBlockSize(ZSTD_CCtx* cctx, const void* src, size_t srcSize, size_t blockSizeMax, ZSTD_strategy strat, S64 savings)
+{
+    /* note: conservatively only split full blocks (128 KB) currently,
+     * and even then only if there is more than 128 KB input remaining.
+     */
+    if (srcSize <= 128 KB || blockSizeMax < 128 KB)
+        return MIN(srcSize, blockSizeMax);
+    /* dynamic splitting has a cpu cost for analysis,
+     * due to that cost it's only used for btlazy2+ strategies */
+    if (strat >= ZSTD_btlazy2)
+        return ZSTD_splitBlock_4k(src, srcSize, blockSizeMax, cctx->tmpWorkspace, cctx->tmpWkspSize);
+    /* blind split strategy
+     * no cpu cost, but can over-split homegeneous data.
+     * heuristic, tested as being "generally better".
+     * do not split incompressible data though: respect the 3 bytes per block overhead limit.
+     */
+    return savings ? 92 KB : 128 KB;
```

Comment:
This looks like it is splitting if the savings is non-zero. So it will split if the savings is negative. Is that correct? If so why do we want to do that?

Reply:
Nope, `savings` will only be `>0` if there were some observed savings (aka, smaller `cSize` than `srcSize`) from previous blocks.

It also means that, for the first block, it will never split.

Reply:
But this currently will also split if the savings is negative, which doesn't seem like the desired behavior. Is this what you intended?

```suggestion
    return savings > 0 ? 92 KB : 128 KB;
```

Reply:
Additionally, we need to compute how much overhead this can possibly add, and ensure that this block split can never grow us beyond the compress bound.

E.g.

```suggestion
    return savings >= 3 ? 92 KB : 128 KB;
```

Reply:
done

Reply:
> But this currently will also split if the savings is negative, which doesn't seem like the desired behavior. Is this what you intended?

That's a very good point.
In my mind, `savings` was boxed to remain positive,
which it is as long as it remains within `compress_frameChunk()`.

But as soon as we leave that function and return (streaming scenario),
`savings` is regenerated from actual bytes consumes and written,
which can be negative if data was so far incompressible.

This is fixed in newer version, thanks to the `>3` test.

Great catch @terrelln !